test/semantics/call12.f90

```diff
@@ -23,58 +27,62 @@ module m
   type :: hasPtr
     real, pointer :: p
   end type
+  type :: hasCoarray
+    real :: co[*]
+  end type
  contains
-  pure subroutine msubr
-    !ERROR: A PURE subprogram must not define a host-associated object
+  pure function test(ptr, in, hpd)
```

Comment:
One of the circumstances that can cause the illegal deallocation of a a polymorphic entity is peculiar to PURE procedures.  It's described in section 9.7.3.2, paragraph 5:

If a function whose result is allocatable or has an allocatable subobject is referenced in the specification part of a scoping unit, and the function reference is executed, an allocatable result and any allocated allocatable subobject of the result is deallocated before execution of the executable constructs of the scoping unit.

I'm not sure if your code covers this case, but it would be good to have a test for it.

Reply:
Doesn't C1584 cover that case?  A pure function can't return a value whose finalization might reference an impure subroutine, and only references to pure functions may appear in specification expressions (10.1.11(4)).

Reply:
Actually, the case I was thinking about is covered by C1585.  Furthermore, this case is already covered by test call10.f90.

So, never mind ...